src/explore.sh

```diff
@@ -2,106 +2,145 @@
-      shift 1
-
-      if [[ "$#" -gt 2 && "$3" != "TEST_MODE" ]]; then
-        complain 'Too many parameters'
-        exit 22 # EINVAL
-      fi
-
-      explore_all_files_git "$@"
-      ;;
-    4) # Search in files under git control
-      if [[ "$#" -gt 2 && "$3" != "TEST_MODE" ]]; then
-        complain 'Too many parameters'
-        exit 22 # EINVAL
-      fi
-
-      explore_files_under_git "$@"
-      ;;
-    *)
-      complain 'Invalid parameter'
-      exit 22 # EINVAL
-      ;;
-  esac
+  if [[ "$1" =~ -h|--help ]]; then
+    explore_help "$1"
+    exit 0
+  fi
+
+  parse_explore_options "$@"
+  if [[ "$?" -gt 0 ]]; then
+    complain "${options_values['ERROR']}"
+    return 22 # EINVAL
+  fi
+
+  flag="${options_values['TEST_MODE']}"
+  search="${options_values['SEARCH']}"
+  path="${options_values['PATH']}"
+
+  if [[ -n "${options_values['GIT_LOG']}" ]]; then
+    # LOG
+    explore_git_log "$search" "$path" "$flag"
+  fi
+
+  if [[ -n "${options_values['GNU_GREP']}" ]]; then
+    # Use GNU GREP
+    explore_files_gnu_grep "$search" "$path" "$flag"
+  fi
+
+  if [[ -n "${options_values['GIT_GREP']}" ]]; then
+    # Search in directories controlled or not by git
+    explore_all_files_git "$search" "$path" "$flag"
+  fi
+
+  if [[ -n "${options_values['FILE_UNDER_GIT']}" ]]; then
+    # Search in files under git control
+    explore_files_under_git "$search" "$path" "$flag"
+  fi
 }
 
 # Tiny parameter parser, we try to keep this function isolate for making easy
 # to expand the explore options.
 #
 # Returns:
 # This function returns the following values:
-#   1 if --log is passed
-#   2 if --grep is passed
-#   3 if --all is passed
-#   4 for something different from --log or --grep
+#   0 if options are parsed successfully
 #   22 for invalid operation
-function explore_parser()
+# This function also set options_values
+function parse_explore_options()
 {
-  local option="$1"
+  local long_options='log,grep,all'
+  local short_options='l,g,a'
+  local options
+  local type_set=0
 
   if [[ "$#" -eq 0 ]]; then
-    complain 'Expected string or parameter. See man for detail.'
-    exit 22 # EINVAL
+    options_values['ERROR']="Expected string or parameter. See man for detail."
+    return 22 # EINVAL
   fi
 
-  if [[ "$1" =~ -h|--help ]]; then
-    explore_help "$1"
-    exit 0
+  options="$(kw_parse "$short_options" "$long_options" "$@")"
+  if [[ "$?" != 0 ]]; then
+    options_values['ERROR']="$(kw_parse_get_errors 'kw explore' \
+      "$short_options" "$long_options" "$@")"
+    return 22 # EINVAL
   fi
 
-  if [[ -z "$2" ]]; then
-    return 4
+  options_values['TEST_MODE']='SILENT'
+  options_values['SEARCH']=''
+  options_values['PATH']=''
+
+  eval "set -- $options"
+
+  while [[ "$#" -gt 0 ]]; do
+    case "$1" in
+      --log | -l)
+        if [[ "$type_set" -eq 1 ]]; then
+          options_values['ERROR']="Invalid arguments: Multiple search type!"
+          return 22 # EINVAL
+        fi
```

Comment:
This code repeats multiple times; let's make it a function (add unit test) and reuse it here. Also, I think `type_set` is not a good name; maybe use something like `is_seach_option_already_set` or `is_it_a_multiple_option_case`?